cardano-crypto-tests/src/Test/Crypto/DSIGN.hs

```diff
@@ -31,166 +136,159 @@ import Test.Tasty.QuickCheck (testProperty)
-      , testProperty "Sig"     $ prop_no_thunks @(SigDSIGN v)
+      ],
+    testGroup "verify" [
+      testProperty "signing and verifying with matching keys" . 
+        forAllShow ((,) <$> genMsg <*> genSK) ppShow $
+        prop_dsign_verify,
+      testProperty "verifying with wrong key" . 
+        forAllShow genWrongKey ppShow $
+        prop_dsign_verify_wrong_key,
+      testProperty "verifying wrong message" . 
+        forAllShow genWrongMsg ppShow $ 
+        prop_dsign_verify_wrong_msg
+      ],
+    testGroup "NoThunks" [
+      testProperty "VerKey" . forAllShow genVK ppShow $ prop_no_thunks,
+      testProperty "SignKey" . forAllShow genSK ppShow $ prop_no_thunks,
+      testProperty "Sig" . forAllShow genSig ppShow $ prop_no_thunks
       ]
     ]
+    where
+      genWrongKey :: Gen (a, SignKeyDSIGN v, SignKeyDSIGN v)
+      genWrongKey = do
+        sk1 <- genSK
+        sk2 <- Gen.suchThat genSK (/= sk1)
+        msg <- genMsg
+        pure (msg, sk1, sk2)
+      genWrongMsg :: Gen (a, a, SignKeyDSIGN v)
+      genWrongMsg = do
+        msg1 <- genMsg
+        msg2 <- Gen.suchThat genMsg (/= msg1)
+        sk <- genSK
+        pure (msg1, msg2, sk)
 
--- | If we sign a message @a@ with the signing key, then we can verify the
--- signature using the corresponding verification key.
---
-prop_dsign_verify_pos
-  :: forall v. (DSIGNAlgorithm v, ContextDSIGN v ~ (), Signable v ~ SignableRepresentation)
-  => Message
-  -> SignKeyDSIGN v
+-- If we sign a message with the key, we can verify the signature with the
+-- corresponding verification key.
+prop_dsign_verify  
+  :: forall (v :: Type) (a :: Type) . 
+  (DSIGNAlgorithm v, ContextDSIGN v ~ (), Signable v a) 
+  => (a, SignKeyDSIGN v)
   -> Property
-prop_dsign_verify_pos a sk =
-  let sig = signDSIGN () a sk
+prop_dsign_verify (msg, sk) = 
+  let signed = signDSIGN () msg sk
       vk = deriveVerKeyDSIGN sk
-  in verifyDSIGN () vk a sig === Right ()
+    in verifyDSIGN () vk msg signed === Right ()
 
--- | If we sign a message @a@ with one signing key, if we try to verify the
--- signature (and message @a@) using a verification key corresponding to a
--- different signing key, then the verification fails.
---
-prop_dsign_verify_neg_key
-  :: forall v. (DSIGNAlgorithm v, Eq (SignKeyDSIGN v),
-                ContextDSIGN v ~ (), Signable v ~ SignableRepresentation)
-  => Message
-  -> SignKeyDSIGN v
-  -> SignKeyDSIGN v
+-- If we sign a message with one key, and try to verify with another, then
+-- verification fails.
+prop_dsign_verify_wrong_key
+  :: forall (v :: Type) (a :: Type) .
+  (DSIGNAlgorithm v, Signable v a, ContextDSIGN v ~ ()) 
+  => (a, SignKeyDSIGN v, SignKeyDSIGN v) 
   -> Property
-prop_dsign_verify_neg_key a sk sk' =
-  sk /= sk' ==>
-    let sig = signDSIGN () a sk
-        vk' = deriveVerKeyDSIGN sk'
-    in verifyDSIGN () vk' a sig =/= Right ()
-
+prop_dsign_verify_wrong_key (msg, sk, sk') = 
```

Comment:
Removing the `sk /= sk'` precondition can lead to false negatives - when the "wrong" key happens to be identical to the correct key, the verification will of course succeed, and the test fails.

Without shrinking, this is very unlikely (by design), but I still think it's prudent to keep the precondition.

Reply:
This precondition is not required, as the generator is now _guaranteed_ not to give equal keys. See `genWrongKey` for how this is done.

Reply:
OK, makes sense.